File: frontend/src/interfaces/Schedule6Response.ts
// Mirrors the backend Schedule 6 (Road Management Costs) DTOs, frozen by Stories 8.1/8.2 — consumed
// here, never re-shaped (AD-12). Jackson omits nulls (non_null), so every nullable member is typed
// `| null` AND may simply be absent from the JSON; read them defensively. `rmg`, `costPerVolume` and
// the three totals are DERIVED server-side (BR-04/BR-07) and are response-only — never recomputed
// here (AD-5) and never sent back.

import type { CodeDescription } from '@/utils/codes'
import type { MessageInfo } from './Schedule1Response'

export type { MessageInfo }

// The TSA-number and Supply-Block dropdown lists, served in-document (Schedule6CodeLists.java).
// Both are year-filtered server-side and always include any code a stored record already references.
// There is no TFL list: TFL stays a free-text entry (schedule6.xhtml:102,290), and the synthetic
// "TFL" sentinel is not a code-table row, so the control — not the document — adds it.
export interface Schedule6CodeLists {
  readonly tsaNumbers: readonly CodeDescription[]
  readonly supplyBlocks: readonly CodeDescription[]
}

// One road-maintenance record. A record is either a Timber Supply Area (`areaType` = the TSA code,
// `supplyBlock` = the TSB code) or a Tree Farm Licence (`areaType` = the literal "TFL", `tflNumber`
// = the TFL code) — BR-02, mutually exclusive, so the counterpart field is always null.
// `revisionCount` is THIS ROW's optimistic-lock token: Schedule 6 has no schedule-level summary row,
// so concurrency is keyed per record (the AR11 delta recorded in Story 8.1).
export interface RoadRecord {
  readonly recordId: number
  readonly revisionCount: number | null
  readonly areaType: string | null
  readonly tflNumber: string | null
  readonly supplyBlock: string | null
  readonly rmg: string | null
  readonly volume: number | null
  // Whole dollars (Integer on the wire).
  readonly cost: number | null
  // Null when volume is zero/absent — 0/0 is undefined, not zero (Story 8.1 deviation (c)).
  readonly costPerVolume: number | null
  // The PER-RECORD comment: ILCR_COST_REPORT_DETAIL.COMMENTS VARCHAR2(400 BYTE), a different and
  // much narrower column than the schedule-level `generalComments` below.
  readonly comments: string | null
}

// One missing-field finding. `field` names the request field the user must supply (`areaType`,
// `tflNumber`, `supplyBlock`, `cost`); `message.text` is the composed verbatim line, e.g.
// "Road : 1 - TSA or TFL (Cost $) : Value Required".
export interface FieldIssue {
  readonly field: string
  readonly message: MessageInfo
}

// One record's Check Status result. `rowCounter` is the 1-based ordinal the user sees in the message
// text (and the accordion title). `recordId` carries that SAME 1-based payload ordinal, not a
// database id — a payload row addresses no stored record, so there is no id to echo; it is a React
// list key only, never a correlation back to a stored row. `metMessage` is present only when this
// record is met AND the schedule outcome is ISSUES, and Jackson may omit it entirely rather than send
// null (Story 8.2 deviation (i)).
export interface RoadRecordCheckResult {
  readonly recordId: number
  readonly rowCounter: number
  readonly met: boolean
  readonly metMessage?: MessageInfo | null
  readonly issues: readonly FieldIssue[]
}

// Check Status (S09–S11, S20, S21) — read-only readiness validation that mutates nothing. A MET
// outcome carries the single schedule banner in `messages` and NO per-record results at all (the
// legacy pass branch never enters the per-record loop).
export interface Schedule6CheckStatusResponse {
  readonly outcome: 'MET' | 'ISSUES'
  readonly messages: readonly MessageInfo[]
  readonly records: readonly RoadRecordCheckResult[]
}

export default interface Schedule6Response {
  readonly millId: number
  readonly year: number
  readonly trackStatus: string | null
  // SERVER-AUTHORITATIVE (AD-9): EDIT_SCHEDULE ∧ trackStatus == "D". Never derived on the client.
  readonly editable: boolean
  // The single schedule-level comment (ROAD_MAINTENANCE_REPORT.COMMENTS, 4000 wide; capped at the
  // legacy UI's 3500). Saved independently of any record (BR-09).
  readonly generalComments: string | null
  // Placeholder rows (a lone general-comment holder) are excluded server-side, so an empty list with
  // a non-empty general comment is the valid S18 state.
  readonly roadRecords: readonly RoadRecord[]
  readonly codeLists: Schedule6CodeLists
  readonly totalVolume: number | null
  readonly totalCost: number | null
  // Null in the S18 lone-comment state (0/0), where totalVolume/totalCost are 0 — renders BLANK.
  readonly totalCostPerVolume: number | null
  // Write echoes only (AD-8); absent on the GET.
  readonly message?: MessageInfo | null
}

File: backend/src/main/java/ca/bc/gov/nrs/ilcr/schedule6/dto/Schedule6Response.java
package ca.bc.gov.nrs.ilcr.schedule6.dto;

import ca.bc.gov.nrs.ilcr.dto.base.MessageInfo;
import com.fasterxml.jackson.annotation.JsonInclude;
import java.math.BigDecimal;
import java.util.List;

/**
 * The Schedule 6 (Road Management Costs) read document (AD-5, AD-12) — the pinned GET response,
 * frozen for all Schedule 6 stories. Like Schedule 4, Schedule 6 is a multi-record structure stored
 * in a dedicated master table ({@code THE.ROAD_MAINTENANCE_REPORT}) with per-record cost/volume in
 * {@code ILCR_COST_REPORT_DETAIL} — NOT the flat {@code ILCR_REPORT_SUMMARY} shape of Sch 1-3, and
 * there is no category-{@code "6"} summary row (delivery-DB confirmed, Story 8.1 Task 1).
 *
 * <p>{@code trackStatus} = {@code ILCR_MILL_REPORT_STATUS_CODE} (the Schedules 1-10 track, AD-9).
 * {@code editable} = the caller holds {@code EDIT_SCHEDULE} AND {@code trackStatus == "D"},
 * computed server-side (AD-5/AD-9), never client-supplied — a non-Draft mill still lists records
 * ({@code editable:false}). {@code generalComments} is the single schedule-level comment (stored
 * replicated on every road-record row in legacy). {@code totalVolume}/{@code totalCost}/ {@code
 * totalCostPerVolume} are DERIVED running totals (BR-07). A valid, active mill/year with no road
 * records returns {@code roadRecords: []} with zero totals and no general comment — never a 404
 * (that is reserved for the mill/year context guard, ERR-003).
 *
 * <p>{@code message} is the AD-8 success-message echo: null on a GET read (Jackson {@code non_null}
 * omits it), carrying the resolved {@link MessageInfo} on the Story 8.2 save echo.
 *
 * <p>{@code codeLists} carries the TSA and Supply Block dropdown options (code + description).
 * Legacy rendered both controls as {@code p:selectOneMenu} with {@code itemLabel} bound to the
 * code's {@code DESCRIPTION} ({@code schedule6.xhtml:265-323}), so the page must be able to show a
 * name for a stored code. This retires deviation (A), which served the raw codes as free text.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record Schedule6Response(
    long millId,
    int year,
    String trackStatus,
    boolean editable,
    String generalComments,
    List<RoadRecord> roadRecords,
    BigDecimal totalVolume,
    Long totalCost,
    BigDecimal totalCostPerVolume,
    Schedule6CodeLists codeLists,
    MessageInfo message) {

  /** A copy of this document carrying the given success message (for the save echo, AD-8). */
  public Schedule6Response withMessage(MessageInfo message) {
    return new Schedule6Response(
        millId,
        year,
        trackStatus,
        editable,
        generalComments,
        roadRecords,
        totalVolume,
        totalCost,
        totalCostPerVolume,
        codeLists,
        message);
  }
}
